File: integration/apiclient/teams.go
package apiclient

// import (
// 	"encoding/json"
// 	"fmt"
// 	"net/http"
// 	"path"
// )

// // These constants are used by various client methods.
// const (
// 	TeamTypeManaged = "managed"
// 	TeamTypeLDAP    = "ldap"
// )

// // Team represents a team within an organization.
// type Team struct {
// 	ID                       uint   `json:"id"`
// 	OrgID                    uint   `json:"orgID"`
// 	Type                     string `json:"type"`
// 	Name                     string `json:"name"`
// 	Description              string `json:"description"`
// 	LdapDN                   string `json:"ldapDN,omitempty"`
// 	LdapGroupMemberAttribute string `json:"ldapGroupMemberAttribute,omitempty"`
// }

// // createTeamForm containst the fields used in CreateTeam requests.
// type createTeamForm struct {
// 	Name                     string `json:"name"`
// 	Description              string `json:"description"`
// 	Type                     string `json:"type"`
// 	LdapDN                   string `json:"ldapDN"`
// 	LdapGroupMemberAttribute string `json:"ldapGroupMemberAttribute"`
// }

// // createTeam creates a team in the given organization with the data in the
// // given form.
// func (c *apiClient) createTeam(orgname string, form createTeamForm) (*Team, error) {
// 	response, err := c.makeRequest("POST", path.Join(accountsBasePath, orgname, "teams"), form)
// 	if err != nil {
// 		return nil, fmt.Errorf("unable to make API request: %s", err)
// 	}
// 	defer response.Body.Close()

// 	if err := validateStatusCode(response, http.StatusCreated); err != nil {
// 		return nil, err
// 	}

// 	var team Team
// 	if err := json.NewDecoder(response.Body).Decode(&team); err != nil {
// 		return nil, fmt.Errorf("unable to decode API response: %s", err)
// 	}

// 	return &team, nil
// }

// // CreateManagedTeam creates a new managed team in an organization with the
// // given teamname and description.
// func (c *apiClient) CreateManagedTeam(orgname, teamname, description string) (*Team, error) {
// 	return c.createTeam(orgname, createTeamForm{
// 		Type:        TeamTypeManaged,
// 		Name:        teamname,
// 		Description: description,
// 	})
// }

// // CreateLDAPGroupSyncedTeam creates a new LDAP group-synced team in an
// // organization with the given teamname, description, groupDN, and memberAttr.
// func (c *apiClient) CreateLDAPGroupSyncedTeam(orgname, teamname, description, groupDN, memberAttr string) (*Team, error) {
// 	return c.createTeam(orgname, createTeamForm{
// 		Type:        TeamTypeLDAP,
// 		Name:        teamname,
// 		Description: description,
// 		LdapDN:      groupDN,
// 		LdapGroupMemberAttribute: memberAttr,
// 	})
// }

// // ListTeams lists all of the teams in an organization.
// func (c *apiClient) ListTeams(orgname string) ([]*Team, error) {
// 	response, err := c.makeRequest("GET", path.Join(accountsBasePath, orgname, "teams"), nil)
// 	if err != nil {
// 		return nil, fmt.Errorf("unable to make API request: %s", err)
// 	}
// 	defer response.Body.Close()

// 	if err := validateStatusCode(response, http.StatusOK); err != nil {
// 		return nil, err
// 	}

// 	var teamsResponse struct {
// 		Teams []*Team `json:"teams"`
// 	}
// 	if err := json.NewDecoder(response.Body).Decode(&teamsResponse); err != nil {
// 		return nil, fmt.Errorf("unable to decode API response: %s", err)
// 	}

// 	return teamsResponse.Teams, nil
// }

// // GetTeam retrieves info about a team in an organization.
// func (c *apiClient) GetTeam(orgname, teamname string) (*Team, error) {
// 	response, err := c.makeRequest("GET", path.Join(accountsBasePath, orgname, "teams", teamname), nil)
// 	if err != nil {
// 		return nil, fmt.Errorf("unable to make API request: %s", err)
// 	}
// 	defer response.Body.Close()

// 	if err := validateStatusCode(response, http.StatusOK); err != nil {
// 		return nil, err
// 	}

// 	var team Team
// 	if err := json.NewDecoder(response.Body).Decode(&team); err != nil {
// 		return nil, fmt.Errorf("unable to decode API response: %s", err)
// 	}

// 	return &team, nil
// }

// // TeamUpdateForm is used to update fields of a team.
// // If Name is not blank, the name will be updated.
// // If Description is not nil, the description will be updated.
// // If LdapDN is not blank and the team is synced with an LDAP group, the LDAP
// // DN of the group will be updated.
// // If LdapGroupMemberAttribute is not blank and the team is synced with an LDAP
// // group, the group member attribute of the team will be updated.
// type TeamUpdateForm struct {
// 	Name                     string  `json:"name"`
// 	Description              *string `json:"description"`
// 	LdapDN                   string  `json:"ldapDN"`
// 	LdapGroupMemberAttribute string  `json:"ldapGroupMemberAttribute"`
// }

// // UpdateTeam updates a team in an organization with the field in the given
// // form.
// func (c *apiClient) UpdateTeam(orgname, teamname string, form TeamUpdateForm) (*Team, error) {
// 	response, err := c.makeRequest("PATCH", path.Join(accountsBasePath, orgname, "teams", teamname), form)
// 	if err != nil {
// 		return nil, fmt.Errorf("unable to make API request: %s", err)
// 	}
// 	defer response.Body.Close()

// 	if err := validateStatusCode(response, http.StatusOK); err != nil {
// 		return nil, err
// 	}

// 	var team Team
// 	if err := json.NewDecoder(response.Body).Decode(&team); err != nil {
// 		return nil, fmt.Errorf("unable to decode API response: %s", err)
// 	}

// 	return &team, nil
// }

// // DeleteTeam deletes a team from an organization.
// // Note: this method may be called repeatedly with no effect.
// func (c *apiClient) DeleteTeam(orgname, teamname string) error {
// 	response, err := c.makeRequest("DELETE", path.Join(accountsBasePath, orgname, "teams", teamname), nil)
// 	if err != nil {
// 		return fmt.Errorf("unable to make API request: %s", err)
// 	}
// 	defer response.Body.Close()

// 	return validateStatusCode(response, http.StatusNoContent)
// }

// // ListTeamMembers lists the users which are a member of a team in an
// // organization.
// func (c *apiClient) ListTeamMembers(orgname, teamname string) ([]*Account, error) {
// 	response, err := c.makeRequest("GET", path.Join(accountsBasePath, orgname, "teams", teamname, "members"), nil)
// 	if err != nil {
// 		return nil, fmt.Errorf("unable to make API request: %s", err)
// 	}
// 	defer response.Body.Close()

// 	if err := validateStatusCode(response, http.StatusOK); err != nil {
// 		return nil, err
// 	}

// 	var membersResponse struct {
// 		Members []*Account `json:"members"`
// 	}
// 	if err := json.NewDecoder(response.Body).Decode(&membersResponse); err != nil {
// 		return nil, fmt.Errorf("unable to decode API response: %s", err)
// 	}

// 	return membersResponse.Members, nil
// }

// // ListOrganizationMemberTeams lists the teams which an organization member
// // belongs to.
// func (c *apiClient) ListOrganizationMemberTeams(orgname, username string) ([]*Team, error) {
// 	response, err := c.makeRequest("GET", path.Join(accountsBasePath, orgname, "members", username, "teams"), nil)
// 	if err != nil {
// 		return nil, fmt.Errorf("unable to make API request: %s", err)
// 	}
// 	defer response.Body.Close()

// 	if err := validateStatusCode(response, http.StatusOK); err != nil {
// 		return nil, err
// 	}

// 	var teamsResponse struct {
// 		Teams []*Team `json:"teams"`
// 	}
// 	if err := json.NewDecoder(response.Body).Decode(&teamsResponse); err != nil {
// 		return nil, fmt.Errorf("unable to decode API response: %s", err)
// 	}

// 	return teamsResponse.Teams, nil
// }

// // ListOrgMembers lists the users which are a member of a team in an
// // organization.
// func (c *apiClient) ListOrgMembers(orgname string) ([]*Account, error) {
// 	response, err := c.makeRequest("GET", path.Join(accountsBasePath, orgname, "members"), nil)
// 	if err != nil {
// 		return nil, fmt.Errorf("unable to make API request: %s", err)
// 	}
// 	defer response.Body.Close()

// 	if err := validateStatusCode(response, http.StatusOK); err != nil {
// 		return nil, err
// 	}

// 	var membersResponse struct {
// 		Members []*Account `json:"members"`
// 	}
// 	if err := json.NewDecoder(response.Body).Decode(&membersResponse); err != nil {
// 		return nil, fmt.Errorf("unable to decode API response: %s", err)
// 	}

// 	return membersResponse.Members, nil
// }

// func (c *apiClient) ListUserOrganizations(username string) ([]*Account, error) {
// 	response, err := c.makeRequest("GET", path.Join(accountsBasePath, username, "organizations"), nil)
// 	if err != nil {
// 		return nil, fmt.Errorf("unable to make API request: %s", err)
// 	}
// 	defer response.Body.Close()

// 	if err := validateStatusCode(response, http.StatusOK); err != nil {
// 		return nil, err
// 	}

// 	var organizationsResponse struct {
// 		Organizations []*Account `json:"organizations"`
// 	}
// 	if err := json.NewDecoder(response.Body).Decode(&organizationsResponse); err != nil {
// 		return nil, fmt.Errorf("unable to decode API response: %s", err)
// 	}

// 	return organizationsResponse.Organizations, nil
// }

// // CheckTeamMembership checks whether a user is a member of a team in an
// // organization.
// func (c *apiClient) CheckTeamMembership(orgname, teamname, username string) (isMember bool, err error) {
// 	response, err := c.makeRequest("GET", path.Join(accountsBasePath, orgname, "teams", teamname, "members", username), nil)
// 	if err != nil {
// 		return false, fmt.Errorf("unable to make API request: %s", err)
// 	}
// 	defer response.Body.Close()

// 	switch response.StatusCode {
// 	case http.StatusNotFound:
// 		// 404 Not Found is returned when the user is not a member
// 		// (or the org, team, or user dose not exist).
// 		return false, nil
// 	case http.StatusNoContent:
// 		// 204 No Content is returned when the user is a member.
// 		return true, nil
// 	}

// 	return false, validateStatusCode(response, 0)
// }

// // CheckOrganizationMembership checks whether a user is a member of an
// // organization.
// func (c apiClient) CheckOrganizationMembership(orgname, username string) (isMember bool, err error) {
// 	response, err := c.makeRequest("GET", path.Join(accountsBasePath, orgname, "members", username), nil)
// 	if err != nil {
// 		return false, fmt.Errorf("unable to make API request: %s", err)
// 	}
// 	defer response.Body.Close()

// 	switch response.StatusCode {
// 	case http.StatusNotFound:
// 		// 404 Not Found is returned when the user is not a member
// 		// (or the org, team, or user dose not exist).
// 		return false, nil
// 	case http.StatusNoContent:
// 		// 204 No Content is returned when the user is a member.
// 		return true, nil
// 	}

// 	return false, validateStatusCode(response, 0)
// }

// // AddTeamMember adds a user to a team within an organization.
// // Note: The team must be a managed (i.e., non-ldap group-synced) team.
// // Note: this method may be called repeatedly with no effect.
// func (c *apiClient) AddTeamMember(orgname, teamname, username string) error {
// 	response, err := c.makeRequest("PUT", path.Join(accountsBasePath, orgname, "teams", teamname, "members", username), nil)
// 	if err != nil {
// 		return fmt.Errorf("unable to make API request: %s", err)
// 	}
// 	defer response.Body.Close()

// 	return validateStatusCode(response, http.StatusOK)
// }

// // DeleteTeamMember removes a user from a team within an organization.
// // Note: The team must be a managed (i.e., non-ldap group-synced) team.
// // Note: this method may be called repeatedly with no effect.
// func (c *apiClient) DeleteTeamMember(orgname, teamname, username string) error {
// 	response, err := c.makeRequest("DELETE", path.Join(accountsBasePath, orgname, "teams", teamname, "members", username), nil)
// 	if err != nil {
// 		return fmt.Errorf("unable to make API request: %s", err)
// 	}
// 	defer response.Body.Close()

// 	return validateStatusCode(response, http.StatusNoContent)
// }

// // DeleteOrganizationMember removes a user from an organization (i.e., all
// // teams in that organization).
// func (c *apiClient) DeleteOrganizationMember(orgname, username string) error {
// 	response, err := c.makeRequest("DELETE", path.Join(accountsBasePath, orgname, "members", username), nil)
// 	if err != nil {
// 		return fmt.Errorf("unable to make API request: %s", err)
// 	}
// 	defer response.Body.Close()

// 	return validateStatusCode(response, http.StatusNoContent)
// }

File: adminserver/ui/src/scripts/actions/teams.js
'use strict';

import _ from 'lodash';
import { Accounts } from 'dtr-js-sdk';
import consts from 'consts';
import { TeamRecord, UserRecord, OrgMemberRecord } from 'records';
import { TeamSchema, UserSchema } from 'normalizers';
import { arrayOf, normalize } from 'normalizr';
import { normalizedToRecords } from 'utils/records';

const AccountsAPI = Accounts.default;

// The fucking enzi api split getting team information into two endpoints:
// one for the team ID, name and description and the other for sync info
// (ldap/managed etc)
//
// here we're just going to combine both APIs and ignore it actionside.
//
// we return objects and make a new team record based off of both responses
export function getTeam(orgName, teamName) {
  const combinator = Promise.all([
    AccountsAPI.getTeam({ name: orgName, team: teamName }).then((resp) => {
      return {
        orgName,
        team: resp.data
      };
    }),
    AccountsAPI.getTeamSyncConfig({ name: orgName, team: teamName }).then((resp) => {
      return {
        orgName,
        team: { name: teamName, ...resp.data }
      };
    })
  ]);

  return {
    type: consts.teams.GET_TEAM,
    meta: {
      promiseScope: [orgName, teamName],
      promise: combinator
    }
  };
}

export function listTeams({ orgName, limit = 100 }) {
  return {
    type: consts.teams.LIST_TEAMS,
    meta: {
      promiseScope: [orgName],
      promise: AccountsAPI.listTeams({ name: orgName }, { limit }).then(
        (resp) => {
          const norm = normalize(resp.data.teams, arrayOf(TeamSchema));
          return {
            orgName,
            data: normalizedToRecords(norm, { 'team': TeamRecord })
          };
        }
      )
    }
  };
}

export function listTeamMembers({ orgName, teamName, limit = 50 }) {
  return {
    type: consts.teams.LIST_MEMBERS,
    meta: {
      promiseScope: [orgName, teamName],
      promise: AccountsAPI
        .listTeamMembers({ name: orgName, team: teamName }, { limit })
        .then( response => {
          const data = normalize(response.data.members, arrayOf(UserSchema));

          return {
            orgName,
            teamName,
            data: normalizedToRecords(data, { 'user': OrgMemberRecord })
          };
        })
    }
  };
}

export function createTeam({ orgName, type, name, ldapDN, ldapGroupMemberAttribute }) {
  return {
    type: consts.teams.CREATE_TEAM,
    meta: {
      promiseScope: [orgName, name],
      promise: AccountsAPI.createTeam({ name: orgName }, {
        type,
        name,
        ldapDN,
        ldapGroupMemberAttribute
      }).then(resp => {
        // after creating a team we must save ldap details, if we need to.
        if (type === 'managed') {
          return {
            orgName,
            team: new TeamRecord(resp.data)
          };
        }

        // format the ldap data into what enzi expects...
        const data = {
          enableSync: true,
          selectGroupMembers: true,
          groupDN: ldapDN,
          groupMemberAttr: ldapGroupMemberAttribute
        };

        return AccountsAPI
          .updateTeamSyncConfig({ name: orgName, team: name }, data)
          .then(syncResp => {
            return {
              orgName,
              team: new TeamRecord({
                ...resp.data,
                ...syncResp.data
              })
            };
          });
      })
    }
  };
}

/**
 * Adds a team within an org
 *
 * TODO: Is this needed?
 */
export function createTeamWithUsers({ orgName, team }, users = []) {
  // TODO:  Improve error handling with N+1 API calls.  If the actual team
  // creation API call fails the team resource has not yet been created.  If
  // adding users to the team fails, the team exists and certain members (who
  // knows which ones...) may not be in the team.
  //
  // The error message should reflect these states.
  //
  // See pending API update regarding adding users to the team at once
  return {
    type: consts.teams.CREATE_TEAM,
    meta: {
      promiseScope: [orgName, team.name],
      promise: AccountsAPI.createTeam({ name: orgName }, team)
        .then( () => {
          // If this is an LDAP team we won't have any users - only add users
          // if this is a managed team and users exist
          if (users.length > 0) {
            let promises = users.map( user => {
              return AccountsAPI.addTeamMember({
                name: orgName,
                team: team.name,
                member: user.name
              });
            });
            // After all members have been added return the new team and all
            // members for the reducer to add to the teams list
            return Promise.all(promises).then( () => {
              return {
                orgName,
                team: new TeamRecord(team),
                members: _.mapValues(_.indexBy(users, 'name'), (member) => new UserRecord(member))
              };
            });
          } else {
            return {
              orgName,
              team: new TeamRecord(team),
              members: {}
            };
          }
        }, error => {
          throw {
            _error: error.data.errors
          };
        }
      ),
      notifications: {
        pending: 'Creating team',
        success: 'Your team was created',
        error: 'There was an error creating your team'
      }
    }
  };
}

export function getTeamsForUser({ name }) {
  return {
    type: consts.teams.GET_TEAMS_FOR_USER,
    meta: {
      promise: AccountsAPI.listUserOrganizations({ name }, { limit: 100 })
        .then((response) => {
          const {
            data: {
              memberOrgs
            }
          } = response;

          // extract the org names from the response
          const orgNames = memberOrgs.map((org) => {
            return org.org.name;
          });

          // get the list of teams for each org
          const promises = orgNames.map((org) => {
            return AccountsAPI.listOrganizationMemberTeams({ name: org, member: name }).then((r) => {
              return r.data.memberTeams.map((team) => {
                team.team.orgName = org;
                return team;
              });
            });
          });

          return Promise.all(promises).then((r) => {
            return {
              user: name,
              teamsByOrg: r
            };
          });

        }, error => {
          throw {
            _error: error.data.errors
          };
        })
    }
  };
}

export function deleteTeam({ orgName, teamName }) {
  return {
    type: consts.teams.DELETE_TEAM,
    meta: {
      promiseScope: [orgName, teamName],
      promise: AccountsAPI.deleteTeam({ name: orgName, team: teamName })
        .then(() => {
          return {
            orgName,
            teamName
          };
        }).then(() => {
          // TODO make this more not suck
          window.location.href = `/orgs/${ orgName }/users`;
        }),
      notifications: {
        pending: 'Deleting team',
        success: `Team ${teamName} was deleted`,
        error: `There was an error deleting ${teamName}`
      }
    }
  };
}

export function updateTeam(orgName, teamName, teamData) {
  // When switching from an LDAP team to managed we need to remove any LDAP
  // fields such as the DN; the API fails if these are sent alongside a managed
  // team.
  // We handle this here so every form need not worry about deleting these
  // fields during submission
  if (teamData.type === 'managed') {
    delete teamData.ldapDN;
    delete teamData.ldapGroupMemberAttribute;
  }

  return {
    type: consts.teams.UPDATE_TEAM,
    meta: {
      promiseScope: [orgName, teamName],
      promise: AccountsAPI.updateTeam({ name: orgName, team: teamName }, teamData)
        .then((resp) => {
          // format the ldap data into what enzi expects...
          const data = {
            enableSync: teamData.type === 'ldap',
            ...teamData,
            selectGroupMembers: true,
            searchBaseDN: '',
            searchScopeSubtree: false,
            searchFilter: ''
          };

          return AccountsAPI
            .updateTeamSyncConfig({ name: orgName, team: resp.data.name }, data)
            .then(syncResp => {
              if (teamName !== resp.data.name) {
                window.location.href = `/orgs/${orgName}/teams/${resp.data.name}/settings`;
              }

              return {
                orgName,
                team: new TeamRecord({
                  ...resp.data,
                  ...syncResp.data
                })
              };
            });
        }),
      notifications: {
        pending: `Updating team ${teamName}`,
        success: `Successfully updated ${teamName}`,
        error: `There was an error updating ${teamName}`
      }
    }
  };
}

export function getTeamSyncOptions({ orgName, teamName }) {
  return {
    type: consts.teams.GET_TEAM_SYNC,
    meta: {
      promiseScope: [orgName, teamName],
      promise: AccountsAPI
        .getTeamSyncConfig({ name: orgName, team: teamName })
    }
  };
}

export function checkTeamMembership(orgName, teamName, memberName) {
  return {
    type: consts.teams.GET_TEAM_MEMBER,
    meta: {
      promiseScope: [orgName, teamName, memberName],
      promise: AccountsAPI
        .checkTeamMembership({ name: orgName, team: teamName, member: memberName })
        .then(
          () => ({
            // This endpoint returns a 204 No Content if the user is
            // within this group.  We only need to return an object for
            // the teams reducer.
            orgName,
            teamName,
            memberName,
            isMember: true
          }),
          (error) => {
            if (error.status === 404 || error.status === 403) {
              // Not found is not an error
              return {
                orgName,
                teamName,
                memberName,
                isMember: false
              };
            }
          }
        )
    }
  };
}

export function addTeamMember({ name, team, member }) {
  return {
    type: consts.teams.ADD_TEAM_MEMBER,
    meta: {
      promiseScope: [name, team],
      promise: AccountsAPI
        .addTeamMember({ name, team, member })
        .then((resp) => ({
          name,
          team,
          member: new UserRecord(resp.data.member)
        }))
    },
    notifications: {
      pending: `Adding ${member} to team ${team}`,
      success: `${member} added to ${team}`,
      error: `There was an error adding ${member} to ${team}`
    }
  };
}

// Add multiple users to a team
export function addTeamMembers(orgName, teamName, memberNames) {
  return {
    type: consts.teams.ADD_TEAM_MEMBERS,
    meta: {
      // TODO:how to scope this? If you call addTeamMembers with a different set of members in succession before the
      // old call can finish, the original call won't fire it's resolved action. This can make our local state's list
      // of members get out of sync.
      // Maybe make sure to block user from adding more members until this is done?
      promiseScope: [orgName, teamName],
      // TODO: If for example one of the add member fails, then Promise.all will fail but the ones which succeeded
      // should still be added to our local state
      promise: Promise.all(
                 memberNames.map((memberName) => {
                   return AccountsAPI.addTeamMember({ name: orgName, team: teamName, member: memberName })
                     .then((resp) => {
                       return resp.data;
                     });
                 })
               ).then((addedMembers) => {
                  const membersArray = addedMembers.map((m) => {
                    return m.member;
                  });
                   return {
                     orgName,
                     teamName,
                     members: _.mapValues(_.keyBy(membersArray, 'name'), (member) => new UserRecord(member))
                   };
               }),
      notifications: {
        pending: `Adding to team ${teamName}`,
        success: `${memberNames.length} new members were added to ${teamName}`,
        error: `There was an error adding members to ${teamName}`
      }
    }
  };
}

// Delete a single member from a team
export function deleteTeamMember({ orgName, memberName, teamName }) {
  return {
    type: consts.teams.DELETE_TEAM_MEMBER,
    meta: {
      promiseScope: [orgName, teamName, memberName],
      promise: AccountsAPI.deleteTeamMember({ name: orgName, team: teamName, member: memberName })
        .then(() => ({
          orgName,
          teamName,
          memberName
        }))
    }
  };
}
